oss-internship-2020/libarchive/examples/sapi_minitar.cc

```diff
@@ -0,0 +1,425 @@
+  std::vector<std::string> relative_paths = absolute_paths;
+
+  std::transform(absolute_paths.begin(), absolute_paths.end(),
+                 absolute_paths.begin(), MakeAbsolutePathAtCWD);
+
+  std::transform(relative_paths.begin(), relative_paths.end(),
+                 relative_paths.begin(), sandbox2::file::CleanPath);
+  // At this point, we have the relative and absolute paths (cleaned) saved
+  // in vectors.
+
+  // Initialize sandbox and api objects.
+  SapiLibarchiveSandboxCreate sandbox(absolute_paths, archive_path);
+  CHECK(sandbox.Init().ok()) << "Error during sandbox initialization";
+  LibarchiveApi api(&sandbox);
+
+  archive* ret = api.archive_write_new().value();
+  CHECK(ret != NULL) << "Failed to create write archive";
+
+  // Treat the pointer as remote. There is no need to copy the data
+  // to the client process.
+  sapi::v::RemotePtr a(ret);
+
+  int ret2;
+
+  switch (compress) {
+    case 'j':
+    case 'y':
+      CHECK(api.archive_write_add_filter_bzip2(&a).value() == ARCHIVE_OK)
+          << "Unexpected result from write_add_filter_bzip2 call";
+      break;
+    case 'Z':
+      CHECK(api.archive_write_add_filter_compress(&a).value() == ARCHIVE_OK)
+          << "Unexpected result from write_add_filter_compress call";
+      break;
+    case 'z':
+      CHECK(api.archive_write_add_filter_gzip(&a).value() == ARCHIVE_OK)
+          << "Unexpected result from write_add_filter_gzip call";
+      break;
+    default:
+      CHECK(api.archive_write_add_filter_none(&a).value() == ARCHIVE_OK)
+          << "Unexpected result from write_add_filter_none call";
+      break;
+  }
+
+  CHECK(api.archive_write_set_format_ustar(&a).value() == ARCHIVE_OK)
+      << "Unexpected result from write_set_format_ustar call";
+
+  const char* filename_ptr = filename.data();
+  if (filename_ptr != NULL && strcmp(filename_ptr, "-") == 0) {
+    filename_ptr = NULL;
+  }
+
+  CHECK(api.archive_write_open_filename(
+               &a, sapi::v::ConstCStr(filename_ptr).PtrBefore())
+            .value() == ARCHIVE_OK)
+      << "Unexpected result from write_open_filename call";
+
+  int file_idx = 0;
+
+  // We can directly use the vectors defined before.
+  for (int file_idx = 0; file_idx < absolute_paths.size(); ++file_idx) {
+    ret = api.archive_read_disk_new().value();
+    CHECK(ret != NULL) << "Failed to create read_disk archive";
+
+    sapi::v::RemotePtr disk(ret);
+
+    CHECK(api.archive_read_disk_set_standard_lookup(&disk).value() ==
+          ARCHIVE_OK)
+        << "Unexpected result from read_disk_set_standard_lookup call";
+
+    // We use the absolute path first.
+    CHECK(
+        api.archive_read_disk_open(
+               &disk,
+               sapi::v::ConstCStr(absolute_paths[file_idx].c_str()).PtrBefore())
+            .value() == ARCHIVE_OK)
+        << CheckStatusAndGetString(api.archive_error_string(&disk), sandbox);
+
+    for (;;) {
+      archive_entry* ret3;
+      ret3 = api.archive_entry_new().value();
```

Comment:
`ret`, `ret2` and `ret3` are slightly confusing - can we use some better name for this?